### PR TITLE
KT-18220 Add data modifier to a class quickfix

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-18220

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.jetbrains.kotlin.idea.search.usagesSearch.descriptor
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.isAbstract
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+
+class AddDataModifierFix(element: KtClass, private val fqName: String) : AddModifierFix(element, KtTokens.DATA_KEYWORD) {
+
+    override fun getText() = "Add data modifier to $fqName"
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: I'd prefer here `Make 'name' data class`

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+            val constructor = callableDescriptor?.returnType?.arguments?.firstOrNull()?.type?.constructor
+                              ?: callableDescriptor?.returnType?.constructor
+
+            val classDescriptor = constructor?.declarationDescriptor as? ClassDescriptor ?: return null
+
+            val klass = DescriptorToSourceUtils.descriptorToDeclaration(classDescriptor) as? KtClass ?: return null
+
+            val ctorParams = klass.primaryConstructor?.valueParameters?.takeIf { it.isNotEmpty() } ?: return null
+            if (!ctorParams.all { it.hasValOrVar() }) return null
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: When some of these val / vars is invisible (e.g. private and we're outside of the class), this fix is not applicable

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                element.getCallableDescriptor()
+
+            val constructor = callableDescriptor?.returnType?.arguments?.firstOrNull()?.type?.constructor
+                              ?: callableDescriptor?.returnType?.constructor
+
+            val classDescriptor = constructor?.declarationDescriptor as? ClassDescriptor ?: return null
+
+            val klass = DescriptorToSourceUtils.descriptorToDeclaration(classDescriptor) as? KtClass ?: return null
+
+            val ctorParams = klass.primaryConstructor?.valueParameters?.takeIf { it.isNotEmpty() } ?: return null
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: I recommend to analyze constructor parameters and modality using descriptor and not PSI. You can convert descriptor to PSI just before quick-fix creation

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt (hunk 2 above)
Comment (mglukhikh) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: Also, data class cannot have vararg parameters.

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt (hunk 1 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: Fixed

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt (hunk 3 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: Fixed

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt (hunk 2 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: Fixed

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            else
+                element.getCallableDescriptor()
+
+            val constructor = callableDescriptor?.returnType?.arguments?.firstOrNull()?.type?.constructor
+                              ?: callableDescriptor?.returnType?.constructor
+
+            val classDescriptor = constructor?.declarationDescriptor as? ClassDescriptor ?: return null
+
+            val modality = classDescriptor.modality
+            if (modality == Modality.ABSTRACT || modality == Modality.SEALED || modality == Modality.OPEN) return null
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: `if (modality != Modality.FINAL)` ...

File: idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            if (ctorParams.isEmpty()) return null
+
+            val isSameContainingClass by lazy {
+                classDescriptor ==  element.getTopmostParentOfType<KtClass>()?.descriptor
+            }
+            val isSameModule by lazy {
+                classDescriptor.module == element.findModuleDescriptor()
+            }
+            if (!ctorParams.all {
+                val param = DescriptorToSourceUtils.descriptorToDeclaration(it) as? KtParameter ?: return@all false
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/AddDataModifierFix.kt]: This code should be also rewritten using descriptors.

**Comment by mglukhikh** [comment]
Fixed and merged manually

**Pull request closed without merging**